### PR TITLE
Issue 41 subsample fn

### LINKED ISSUE #41: function for generating subsample of sample IDs for assays
Several sample IDs are generated in the field sheets. The function needs to:

- pull sample IDs that were used in the field
- identify a subset of those sample IDs according to logic sent from Sean:https://docs.google.com/document/d/1aNT4vqLWFA_4bNRnO9ZemTOTR9L3fyCe/edit?usp=drive_link&ouid=105268335077759969705&rtpof=true&sd=true (updated 11-2023)

### PR DESCRIPTION
**Pull request by liz-fw**
closes #41 

you can test this function by running the code:

`subsamples <- generate_subsample(con, 2024)`

I tried to document the logic as well as I could but let me know if you want to walk through it together to confirm it's working as expected. Logic is here: https://docs.google.com/document/d/1aNT4vqLWFA_4bNRnO9ZemTOTR9L3fyCe/edit

### CODE REVIEW

File: R/generate_subsample.R [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    cli::cli_abort("Season must be a numeric value.")
+  }
+  if(nchar(season) != 4) {
+    cli::cli_abort("Season must be in the format YYYY (i.e. 2024).")
+  }
+  # get all sample event IDs for a season
+  samples <- grunID::filter_dataset(con, season) |>
+    dplyr::add_count(location_code, sample_event_number, name = "no_samples") |>
+    dplyr::add_count(location_code, sample_event_number, sample_bin_code, name = "no_samples_per_bin")
+  dplyr::mutate(no_samples = count(location_code, sample_number),
```
Comment (ergz) [inline R/generate_subsample.R:41]: error here, I think maybe missing a `|>` ?

**Review by ergz — CHANGES_REQUESTED** [review]
I get an error trying to run the function 

```
Error in generate_subsample(con, 2024) : object 'location_code' not found
```

**Comment by liz-fw** [comment]
@ergz those were lines of code i was testing out but we didn't need! should be working now

File: R/generate_subsample.R [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#' @md
+generate_subsample <- function(con, season) {
+  if(!is.numeric(season)) {
+    cli::cli_abort("Season must be a numeric value.")
+  }
+  if(nchar(season) != 4) {
+    cli::cli_abort("Season must be in the format YYYY (i.e. 2024).")
+  }
+  # get all sample event IDs for a season
+  samples <- grunID::filter_dataset(con, season) |>
```
Comment (ergz) [inline R/generate_subsample.R:38]: now that `grunID::filter_dataset` is being used in more places in the code base we should rename it to something more specific like `sample_filter_to_season` or something like that, just to make it clear what it returns

File: R/generate_subsample.R [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    dplyr::add_count(location_code, sample_event_number, name = "no_samples") |>
```
Comment (ergz) [inline R/generate_subsample.R:39]: Instead of `name = "no_samples"` I would use `name = "total_samples_in_event"`

**Review by ergz — APPROVED** [review]
good stuff! I learned about `group_split` and `add_count` now 😄. Logic makes sense, some nip pick stuff for if you have time

1. generate user friendly error when season is not valid
2. after generating subsample produce a summary, I am thinking like what events were sampled 100% and which were chosen at random. This can be a new "feature" issue

**Pull request merged**